File: elixir/secret-handshake/lib/secret_handshake.ex
defmodule SecretHandshake do
  @doc """
  Determine the actions of a secret handshake based on the binary
  representation of the given `code`.

  If the following bits are set, include the corresponding action in your list
  of commands, in order from lowest to highest.

  1 = wink
  10 = double blink
  100 = close your eyes
  1000 = jump

  10000 = Reverse the order of the operations in the secret handshake
  """
  @spec commands(code :: integer) :: list(String.t())
  def commands(code) do
    digits = code |> Integer.digits(2) |> Enum.slice(-5, 5) |> Enum.reverse()
    digits
    |> Enum.with_index()
    |> Enum.filter(fn {d, _} -> d == 1 end)
    |> Enum.map(fn {_, p} -> Integer.pow(2, p) end)
    |> Enum.reduce([], &exec/2)
  end

  @spec exec(command :: integer, state :: list(String.t())) :: list(String.t())
  defp exec(command, state)
  defp exec(1, state), do: Enum.concat(state, ["wink"])
  defp exec(2, state), do: Enum.concat(state, ["double blink"])
  defp exec(4, state), do: Enum.concat(state, ["close your eyes"])
  defp exec(8, state), do: Enum.concat(state, ["jump"])
  defp exec(16, state), do: Enum.reverse(state)
end

File: cpp/secret-handshake/secret_handshake.cpp
#include "secret_handshake.h"

namespace secret_handshake {

using action_and_code = std::pair<std::string, int>;

bool should_do_action(action_and_code action, int code) {
  return (action.second & code);
}

std::vector<std::string> commands(int code) {
  const std::vector<action_and_code> normal_commands{
      {"wink", 1}, {"double blink", 2}, {"close your eyes", 4}, {"jump", 8}};
  const action_and_code reverse_order = {"reverse handshake commands order", 16};
  std::vector<std::string> commands;

  for (action_and_code action : normal_commands) {
    if (should_do_action(action, code)) {
      commands.push_back(action.first);
    }
  }

  if (should_do_action(reverse_order, code)) {
    std::reverse(commands.begin(), commands.end());
  }
  return commands;
}

} // namespace secret_handshake
